### ISSUE TITLE
HikariCP seems to ignore Spring Cloud Connectors' PoolConfig and ConnectionConfig 

### ISSUE THREAD

**Issue by balteo** [spring-cloud/spring-cloud-connectors#173]
I am not 100% certain this is a bug but I have an issue with my use of SCC relational service datasource configuration especially the PoolConfig and ConnectionConfig.

Here is my SCC datasource configuration:

```
@Configuration
@Profile({Profiles.CLOUD, Profiles.DEV})
public class DataSourceConfiguration extends AbstractCloudConfig {

    @Bean
    @Qualifier("application_database")
    @Primary
    public DataSource applicationDatasource() {
        PoolConfig poolConfig = new PoolConfig(5, 11, 3000);//Notice 11 for maxPoolSize
        ConnectionConfig connConfig = new ConnectionConfig("poolName=dummy;characterEncoding=UTF-8");
        DataSourceConfig dbConfig = new DataSourceConfig(poolConfig, connConfig);
        return connectionFactory().dataSource("application_database", dbConfig);
    }
}
```

One will notice the `maxPoolSize` of **11** (I chose **11** so to stand out in the HikariCP report).

However, the HikariCP report seems to indicate that my `PoolConfig` and `ConnectionConfig` are ignored (notice the `maximumPoolSize` set to **10**). See below:

```
16:56:55.916 [main] INFO  o.s.c.s.r.PooledDataSourceCreator - Found HikariCP on the classpath. Using it for DataSource connection pooling.
16:56:56.315 [main] DEBUG com.zaxxer.hikari.HikariConfig - HikariPool-1 - configuration:
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - allowPoolSuspension.............false
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - autoCommit......................true
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - catalog.........................null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - connectionInitSql...............null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - connectionTestQuery............."SELECT 1"
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - connectionTimeout...............30000
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - dataSource......................null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - dataSourceClassName.............null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - dataSourceJNDI..................null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - dataSourceProperties............{password=<masked>}
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - driverClassName................."org.postgresql.Driver"
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - healthCheckProperties...........{}
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - healthCheckRegistry.............null
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - idleTimeout.....................600000
16:56:56.318 [main] DEBUG com.zaxxer.hikari.HikariConfig - initializationFailFast..........true
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - isolateInternalQueries..........false
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - jdbc4ConnectionTest.............false
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - jdbcUrl........................."jdbc:postgresql://localhost:5432/bignibou_application?user=postgres&password=postgres"
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - leakDetectionThreshold..........0
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - maxLifetime.....................1800000
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - maximumPoolSize.................10
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - metricRegistry..................null
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - metricsTrackerFactory...........null
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - minimumIdle.....................10
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - password........................<masked>
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - poolName........................"HikariPool-1"
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - readOnly........................false
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - registerMbeans..................false
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - scheduledExecutorService........null
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - threadFactory...................null
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - transactionIsolation............null
16:56:56.319 [main] DEBUG com.zaxxer.hikari.HikariConfig - username........................null
16:56:56.320 [main] DEBUG com.zaxxer.hikari.HikariConfig - validationTimeout...............5000
16:56:56.320 [main] INFO  com.zaxxer.hikari.HikariDataSource - HikariPool-1 - Started.
16:56:56.387 [main] INFO  com.zaxxer.hikari.pool.PoolBase - HikariPool-1 - Driver does not support get/set network timeout for connections. (La fonction org.postgresql.jdbc.PgConnection.getNetworkTimeout() n'est pas encore implémentée.)
16:56:56.411 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@57a3cabf
16:56:56.505 [HikariPool-1 housekeeper] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Pool stats (total=1, active=1, idle=0, waiting=0)
16:56:56.513 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@3a34aac7
16:56:56.541 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@74fe29ac
16:56:56.551 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@3772d6e0
16:56:56.557 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@7ecd1483
16:56:56.563 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@6130eb0f
16:56:56.568 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@64860dcc
16:56:56.573 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@45e2026a
16:56:56.581 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@539d6f59
16:56:56.586 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - Added connection org.postgresql.jdbc.PgConnection@35939eb6
16:56:56.586 [HikariPool-1 connection adder] DEBUG com.zaxxer.hikari.pool.HikariPool - HikariPool-1 - After adding stats (total=10, active=1, idle=9, waiting=0)
```

Please let me know if this is an issue and how to circumvent it?

P.S. I have opened a thread on SO here: http://stackoverflow.com/questions/38662036/

**Comment by scottfrederick** [comment]
It looks like you have indeed uncovered a bug. The `PoolConfig` class attempts to [map the pool parameters to the appropriate bean property names](https://github.com/spring-cloud/spring-cloud-connectors/blob/fd1677851fa7239a7803aded062c48858f88d5f2/spring-cloud-spring-service-connector/src/main/java/org/springframework/cloud/service/PooledServiceConnectorConfig.java#L44-L86) for each of the supported pooled data source providers. The `maxPoolSize` parameter in `PoolConfig` needs to be mapped to `maximumPoolSize` for HikariCP, but there isn't a getter with this name.

**Comment by balteo** [comment]
Then how are HikariCP-specific properties supposed to be picked up by Spring Cloud Connectors? 

Have a look at my attempt to modify the `poolName` (HikariCP-specific property): `poolName=dummy`. 

The HikariCP reports `poolName........................"HikariPool-1"` indicating my overriding attempt was not picked up either by SCC.

**Comment by balteo** [comment]
Using the following constructor from DataSourceConfig:

```
public DataSourceConfig(PoolConfig poolConfig, ConnectionConfig connectionConfig,
							List<String> pooledDataSourceNames, Map<String, Object> properties)
```

one can pass an Map<String, Object> that is going to be used to instantiate a `private final MapServiceConnectorConfig connectionProperties;` from DataSourceConfig.

See example below that allows to define a HikariCP pool name:
```
@Configuration
@Profile({Profiles.CLOUD, Profiles.DEV})
public class DataSourceConfiguration extends AbstractCloudConfig {

    @Bean
    @Qualifier("application_database")
    @Primary
    public DataSource applicationDatasource() {
        PoolConfig poolConfig = new PoolConfig(5, 11, 3000);
        Map<String, Object> connProperties = new HashMap<>();
        connProperties.put("poolName", "myPool");//This is going to be picked up properly
        DataSourceConfig dbConfig = new DataSourceConfig(poolConfig, null, null, connProperties);
        return connectionFactory().dataSource("application_database", dbConfig);
    }
}

```

Thus one can pass any arbitrary properties including HikariCP's **poolName** and **maximumPoolSize**.

This is a workaround.

**Comment by balteo** [comment]
Or more simply using the **one-arg constructor** as follows:

```
@Configuration
@Profile({Profiles.CLOUD, Profiles.DEV})
public class DataSourceConfiguration extends AbstractCloudConfig {

    @Bean
    @Qualifier("application_database")
    @Primary
    public DataSource applicationDatasource() {
        Map<String, Object> connProperties = new HashMap<>();
        connProperties.put("poolName", "myPool");
        DataSourceConfig dbConfig = new DataSourceConfig(connProperties);
        return connectionFactory().dataSource("application_database", dbConfig);
    }
}
```

**Comment by scottfrederick** [comment]
The map of connection properties was meant to solve the problem of each type concrete connection implementation providing its own set of bean properties that it would be difficult for Connectors to normalize. Setting the `poolName` property fits into this use case nicely. The pool size should, of course, be handled correctly by `PoolConfig`. 

I'm glad you have a work-around for now, we'll keep these issue open to address the pool size mapping problem.

**Comment by balteo** [comment]
> The pool size should, of course, be handled correctly by PoolConfig.

I see what you mean. It should indeed.

**Comment by scottfrederick** [comment]
Fixed via 7d090c999347e86cd5b602f30f03fd90601f2ca0
